### PR TITLE
ec: General cleanup

### PR DESCRIPTION
**Pull request by leoetlino**
This PR tries to improve readability and clarity of the ec code:

- Move all of the ec functions into the Common::ec namespace.

- Give the public functions better names and some usage information.

- Move all of the "elt" related functions into an "elt" class including all of the arithmetic operations, so that the logic becomes clearer and feels less like assembly.

  This also makes it much more obvious what the parameters are, instead of only using unsigned char* (which doesn't tell anything about what the pointer is used for or the size).

- Similarly, add a new "Point" class and move point functions there. Overload the arithmetic operators to make calculations easier to read.

The changeset is unfortunately a bit larger than I'd have liked it to be, but splitting it into smaller commits would have resulted in much more extra work.

[pywii has a variant of the original ec code](https://github.com/fail0verflow/hbc/blob/2fb61e432eed2fffb9f24a089067a7c2ec95fb1f/pywii/Common/pywii/ec.py#L50) that is easier to read, and it's what I used to check for mistakes during the cleanup.

I've also added unit tests for ec functions (to check for mistakes). All tests pass so there shouldn't be any 
 regression.

### CODE REVIEW

File: Source/Core/Common/Crypto/bn.cpp [hunk 1]
```diff
@@ -8,21 +8,19 @@
...
 }
 
-static void bn_copy(u8* d, const u8* a, u32 n)
+static void bn_copy(u8* d, const u8* a, std::size_t n)
 {
   memcpy(d, a, n);
 }
 
-int bn_compare(const u8* a, const u8* b, u32 n)
+int bn_compare(const u8* a, const u8* b, std::size_t n)
```
Comment (Warepire) [inline Source/Core/Common/Crypto/bn.cpp]: Can't this entire function be collapsed into a clamped `memcmp`?

File: Source/Core/Common/Crypto/ec.cpp [hunk 2]
```diff
@@ -20,314 +21,213 @@
...
-  itoh_tsujii(s, t, a, 1);
-  itoh_tsujii(t, s, s, 7);
-  itoh_tsujii(s, t, t, 14);
-  itoh_tsujii(t, s, a, 1);
-  itoh_tsujii(s, t, t, 29);
-  itoh_tsujii(t, s, s, 58);
-  itoh_tsujii(s, t, t, 116);
-  elt_square(d, s);
+  Elt d;
+  for (size_t i = 0; i < std::tuple_size<decltype(Elt::data)>{}; i++)
```
Comment (Warepire) [inline Source/Core/Common/Crypto/ec.cpp]: `std::size_t`?

File: Source/Core/Common/Crypto/ec.cpp [hunk 3]
```diff
@@ -20,314 +21,213 @@
...
-  elt_square_to_wide(wide, a);
-  wide_reduce(wide);
-
-  elt_copy(d, wide + 30);
-}
-
-static void itoh_tsujii(u8* d, const u8* a, const u8* b, u32 j)
-{
-  u8 t[30];
+  Elt ItohTsujii(const Elt& b, size_t j) const
```
[the remark this answers is not in the document]
Comment (leoetlino) [reply Source/Core/Common/Crypto/ec.cpp]: For consistency with bn, I've changed ec to use std::size_t as well.

File: Source/Core/Common/Crypto/bn.cpp (hunk 1 above)
Comment (leoetlino) [reply Source/Core/Common/Crypto/bn.cpp]: Done. (my original goal was to clean up ec, not bn :P)

File: Source/Core/Common/Crypto/bn.h [hunk 4]
```diff
@@ -8,9 +8,9 @@
 
 // bignum arithmetic
 
-int bn_compare(const u8* a, const u8* b, u32 n);
-void bn_sub_modulus(u8* a, const u8* N, u32 n);
-void bn_add(u8* d, const u8* a, const u8* b, const u8* N, u32 n);
-void bn_mul(u8* d, const u8* a, const u8* b, const u8* N, u32 n);
-void bn_inv(u8* d, const u8* a, const u8* N, u32 n);  // only for prime N
-void bn_exp(u8* d, const u8* a, const u8* N, u32 n, const u8* e, u32 en);
+int bn_compare(const u8* a, const u8* b, std::size_t n);
```
[the remark this answers is not in the document]
Comment (leoetlino) [reply Source/Core/Common/Crypto/bn.h]: You're right, it doesn't (I thought it would). `<cstddef>` is now included.

**Comment by Tilka** [comment]
Would it be possible to use mbedtls for this?

**Comment by leoetlino** [comment]
Unfortunately no. mbedtls does not support ECC-B233 (which is the curve used by IOS). We would have to replace mbedtls with openssl if we wanted to use a crypto library.

**Comment by leoetlino** [comment]
Signing seems to be broken. Investigating.

**Comment by leoetlino** [comment]
Fixed a regression that was introduced by the bn loop commit (silly off-by-one).

Also added yet another unit test to verify that signatures are valid.

**Comment by akien-mga** [comment]
Just noting it down before looking further, the new `Common::ec` namespace introduced in 355b1b5 seems to cause a build issue on my system with GCC 5.5.0:
```
[  5%] Building CXX object Source/Core/Common/CMakeFiles/common.dir/Crypto/ec.cpp.o
cd /home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/build/Source/Core/Common && /bin/ccache /bin/c++  -DDATA_DIR=\"/usr/share/dolphin-emu/\" -DHAVE_FFMPEG -DHAVE_LIBEVDEV=1 -$
In file included from /home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:17:0:
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.h:11:17: error: expected '{' before '::' token
 namespace Common::ec
                 ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.h:11:19: error: 'ec' in namespace '::' does not name a type
 namespace Common::ec
                   ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:24:17: error: expected '{' before '::' token
 namespace Common::ec
                 ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:24:19: error: 'ec' in namespace '::' does not name a type
 namespace Common::ec
                   ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:321:1: error: expected '}' at end of input
 }  // namespace Common::ec
 ^
/home/akien/Mageia/Checkout/dolphin-emu/BUILD/dolphin-5.0+7862/Source/Core/Common/Crypto/ec.cpp:321:1: error: expected '}' at end of input
Source/Core/Common/CMakeFiles/common.dir/build.make:281: recipe for target 'Source/Core/Common/CMakeFiles/common.dir/Crypto/ec.cpp.o' failed
```
It builds fine with GCC 7.3.0.

**Comment by akien-mga** [comment]
Looks like nested namespace definitions are a [C++17 feature](http://www.nuonsoft.com/blog/2017/08/01/c17-nested-namespaces/) which is [only available in GCC 6.0 and later](https://gcc.gnu.org/projects/cxx-status.html#cxx17).

So you'd have to either:
- up the minimal GCC version required in https://github.com/dolphin-emu/dolphin/blob/master/CMakeLists.txt#L206
- or avoid using nested namespaces for now, using instead `namespace Common { namespace ec { } }`

**Pull request merged**